sysdig/resource_sysdig_secure_cloud_auth_account.go

```diff
@@ -608,157 +498,86 @@ func featureToResourceData(features *cloudauth.AccountFeatures) []interface{} {
+					},
+				})
+				if err != nil {
+					diag.FromErr(err)
 				}
+				resourceData[SchemaServicePrincipalMetadata] = string(spGcpBytes)
+			} else {
+				resourceData[SchemaServicePrincipalMetadata] = getComponentMetadataString(component.GetServicePrincipalMetadata())
 			}
+		case cloudauth.Component_COMPONENT_WEBHOOK_DATASOURCE:
+			resourceData[SchemaWebhookDatasourceMetadata] = getComponentMetadataString(component.GetWebhookDatasourceMetadata())
+		case cloudauth.Component_COMPONENT_CRYPTO_KEY:
+			resourceData[SchemaCryptoKeyMetadata] = getComponentMetadataString(component.GetCryptoKeyMetadata())
+		case cloudauth.Component_COMPONENT_CLOUD_LOGS:
+			resourceData[SchemaCloudLogsMetadata] = getComponentMetadataString(component.GetCloudLogsMetadata())
 		}
-
-		allComponents[comp.Instance] = componentBlock
+		resourceList = append(resourceList, resourceData)
 	}
 
-	// In the resource data, SchemaComponent field is a list of component sets[] / block
-	// Hence we need to return this uber level list in same order to cloudauthAccountToResourceData
-	componentsList := []map[string]interface{}{}
-
-	if len(allComponents) > 0 {
-
-		// add the component blocks in same order to maintain ordering
-		// note-request; why is this needed?
-		if len(dataComponentsOrder) > 0 {
-			for _, c := range dataComponentsOrder {
-				componentItem := allComponents[c].(map[string]interface{})
-				componentsList = append(componentsList, componentItem)
-			}
-			return componentsList
-		}
-
-		// if no ordering is provided, return all components in any order
-		for _, c := range allComponents {
-			componentItem := c.(map[string]interface{})
-			componentsList = append(componentsList, componentItem)
-		}
-		return componentsList
-
-	}
+	return resourceList
+}
 
-	return nil
+// internal type redefintion for GCP service principals.
+// This exists because in terraform, the key is originally provided in the form of a base64 encoded json string
+type internalServicePrincipalMetadata_GCP struct {
+	Key                        string                                                             `json:"key,omitempty"` // base64 encoded
+	WorkloadIdentityFederation *cloudauth.ServicePrincipalMetadata_GCP_WorkloadIdentityFederation `json:"workload_identity_federation,omitempty"`
+	Email                      string                                                             `json:"email,omitempty"`
+}
+type internalServicePrincipalMetadata struct {
+	Gcp *internalServicePrincipalMetadata_GCP `json:"gcp,omitempty"`
 }
 
-func getResourceComponentsOrder(dataComponents interface{}) []string {
-	var dataComponentsOrder []string
-	for _, rc := range dataComponents.([]interface{}) {
-		resourceComponent := rc.(map[string]interface{})
-		dataComponentsOrder = append(dataComponentsOrder, resourceComponent[SchemaInstance].(string))
+func getComponentMetadataString(message protoreflect.ProtoMessage) string {
+	// marshal through protojson get correct snake case keys
```

Comment:
protojson would give us camelCase from cloudauth. And we re-marshal that using json to give us snake_case to match TF, correct?

Reply:
protojson can do either camel or snake, and invoke it for snake. the reason we're using both is because the key order that encoding/json uses some kind of key ordering that agrees with terraform marshalling.

we use protojson at all because it makes the code far cleaner.